library/BrazeEngineUserJourneyHandler.m

```diff
@@ -0,0 +1,167 @@
+  if (self = [super init]) {
+    _maxAttachedPlacesPerEvent = maxAttachedPlaceEventsPerEvent;
+  }
+  
+  return self;
+}
+
+- (id) init {
+  return [self initWithMaxAttachedPlaceEventsPerEvent:20];
+}
+
+- (void)userJourneyEventDidOccur:(nonnull UserJourneyEvent *)userJourneyEvent { /* Not supported */ }
+
+- (void)userJourneySpanDidOccur:(nonnull UserJourneySpan *)userJourneySpan {
+  // Only send span data packet when did travel is false and tracking User Journey Spans is enabled
+  if (![BrazeEngine isTrackingSpans] || userJourneySpan.didTravel) {
+    return;
+  }
+  
+  // Span information
+  NSString *spanId = userJourneySpan.spanId;
+  BOOL startTimestampUnavailable = userJourneySpan.startTimestampUnavailable;
+  BOOL endTimestampUnavailable = userJourneySpan.endTimestampUnavailable;
+  UInt64 startTimestamp = userJourneySpan.startTimestamp.timeIntervalSince1970 * 1000;
+  UInt64 endTimestamp = userJourneySpan.endTimestamp.timeIntervalSince1970 * 1000;
+  
+  // Current place information
+  FactualPlaceVisit *currentPlace = userJourneySpan.currentPlace;
+  NSNumber *isHome = [[NSNumber alloc] initWithBool:currentPlace.isHome];
+  NSNumber *isWork = [[NSNumber alloc] initWithBool:currentPlace.isWork];
+  
+  // Ingress information
+  CLLocation *ingressLocation = currentPlace.ingressLocation;
+  NSNumber *ingressLatitude = @(ingressLocation.coordinate.latitude);
+  NSNumber *ingressLongitude = @(ingressLocation.coordinate.longitude);
+  
+  // Geographies information
+  Geographies *geographies = currentPlace.geographies;
+  NSString *country = geographies.country;
+  NSArray<NSString*> *localities = geographies.localities;
+  NSString *postcode = geographies.postcode;
+  NSString *region = geographies.region;
+  
+  // Get duration of span
+  NSNumber *duration = !startTimestampUnavailable && !endTimestampUnavailable ? @(endTimestamp - startTimestamp) : 0;
+  
+  // Populate properties data
+  NSMutableDictionary *properties = [[NSMutableDictionary alloc] init];
+  [properties setValue:spanId forKey:SPAN_ID_KEY];
+  [properties setValue:SPAN_SOURCE_NAME forKey:SPAN_SOURCE_KEY];
+  [properties setValue:@(startTimestampUnavailable) forKey:START_TIMESTAMP_UNAVAILABLE_KEY];
+  [properties setValue:@(endTimestampUnavailable) forKey:END_TIMESTAMP_UNAVAILABLE_KEY];
+  [properties setValue:@(startTimestamp) forKey:START_TIMESTAMP_KEY];
+  [properties setValue:@(endTimestamp) forKey:END_TIMESTAM_KEY];
+  [properties setValue:duration forKey:DURATION_KEY];
+  [properties setValue:isHome forKey:IS_HOME_KEY];
+  [properties setValue:isWork forKey:IS_WORK_KEY];
+  [properties setValue:ingressLatitude forKey:INGRESS_LATITUDE_KEY];
+  [properties setValue:ingressLongitude forKey:INGRESS_LONGITUDE_KEY];
+  [properties setValue:country forKey:COUNTRY_KEY];
+  [properties setValue:localities forKey:LOCALITIES_KEY];
+  [properties setValue:postcode forKey:POSTCODE_KEY];
+  [properties setValue:region forKey:REGION_KEY];
```

Comment:
Why not use `[userJourneySpan toDict]`?

Reply:
Because many of the required data points are nested in other objects that cannot be accessed from Braze's dashboard.

Reply:
@nnlamb25 cc @artstwolf our `toDict` methods are supposed to recursively convert nested objects according to their own `toDict` methods. Can you give an example of where this falls short here? Also, if there are deficiencies in `toDict` for a use case like this (which to my mind seems like a solid candidate for why we provide `toDict` in the first place), maybe we should open tickets to address them.

Reply:
Discussed with @nnlamb25 offline. The nesting issue is that Braze needs sub-object data to be flattened into collections at the top level. Therefore structuring the dictionary as a tree of nested objects (as `toDict` does) is not compatible with Braze. Marking this as resolved.